### DISCUSSION TITLE
Migration script now available

### DISCUSSION THREAD

**Discussion by susodapop** [getredash/redash-toolbelt#68] [Announcements]
Hi everyone, we just published version 0.1.5 to Pypi which includes the `redash-migrate` example script 🎉 

Many thanks to our beta testers including: @JSpenced, @ozgenbaris1, @bmtKIA6, @surgachsurgach, and @ChloeConnor.

You can find instructions for using the script in  `docs/redash-migrate/README.md`. Or just click [here](https://github.com/getredash/redash-toolbelt/tree/master/redash_toolbelt/docs/redash-migrate). 

If you encounter issues running the script, please create a new discussion here on Github.

**Comment by roni-frantchi** [comment]
Thanks @susodapop .  
You've mentioned in [this comment](https://github.com/getredash/redash-toolbelt/pull/23#issuecomment-929226924) AMI versions that match the hosted version would be released alongside the migration script, but I'm only [seeing V8 still](https://redash.io/help/open-source/setup#aws) - should I look elsewhere?

**Reply by susodapop** [reply]
You're right that the V10 AMIs aren't up yet. There are a couple follow-ups for the migration script (#64 and #66 ) to handle before we focus on the new cloud images. We'll update [this page](https://redash.io/help/open-source/setup) in the docs once the new images are available.

Until then, you can deploy a V8 AMI and upgrade it to the V10 docker image using the standard upgrade instructions (reproduced here):


### How to upgrade

1. Make sure to backup your data. You only need to backup Redash’s PostgreSQL database (the database Redash stores metadata in, not the ones you might be querying) as the data in Redis is transient (you can skip this step if this is a new instance).
2. Change directory to `/opt/redash`.
3. Modify `/opt/redash/docker-compose.yml` as follows:
  - Set the image reference to the one you want to upgrade to. For V10 this is `redash/redash:10.0.0.b50363`
  - Under `services`/`scheduler`/`environment`, omit `QUEUES` and `WORKERS_COUNT` (and omit `environment` altogether if it is empty).
  - Under services, add a new service for general RQ jobs:

  ```yaml
  worker:
    <<: *redash-service
    command: worker
    environment:
      QUEUES: "periodic emails default"
      WORKERS_COUNT: 1
  ```

4. Stop Redash services: `docker-compose stop server scheduler scheduled_worker adhoc_worker` (you might need to list additional services if you updated your configuration)
5. Apply database migrations: `docker-compose run --rm server manage db upgrade`
6. Force recreation of the contains: `docker-compose up --force-recreate --build`

**Reply by wtfzambo** [reply]
Yo, thanks for the reply.

Just to make sure everything is linear: in the releases page [here](https://github.com/getredash/redash/releases), you mention that aside from bumping the version from 8 to 10 in `docker-compose.yml`, one should also add the following lines:

```yaml
worker:
  <<: *redash-service
  command: worker
  environment:
    QUEUES: "periodic emails default"
    WORKERS_COUNT: 1
```

Does this apply also for the situation you described above ?

**Reply by susodapop** [reply]
Great catch. I edited my comment with the complete instruction set. Thanks!

**Reply by roni-frantchi** [reply]
Thanks @susodapop - just to confirm - I won't be able to use the migration script till I upgrade to V10, is that correct?

**Reply by susodapop** [reply]
That's correct. This script is intended for migrating from app.redash.io to Open Source V10. It may work in other cases but we haven't tested that.

**Reply by wtfzambo** [reply]
@susodapop hey, can you please include to the release page also the fact that one is supposed to run the migration by using this command?

`docker-compose run --rm server manage db upgrade`

I was completely unaware of that and I kept banging my head on the desk for like 4 hours until I figured that out.

**Reply by susodapop** [reply]
Done @wtfzambo . Good lookin' out.

**Reply by roni-frantchi** [reply]
> There are a couple follow-ups for the migration script (#64 and #66 ) to handle before we focus on the new cloud images.

Hey @susodapop. 
I see that both #64 and #66 are now closed.  
Does that mean we can now expect the new AMI to roll out soon?
Thanks!

**Comment by wtfzambo** [comment]
Following

**Comment by kaz3284** [comment]
@susodapop Hi, first of all thanks for your greate efforts!
I had trouble on redash-migration command「groups」. Which it made duplicate 「admin」「default」. 
users looks like divided in eath duplicate groups.
Is it bug?  or. had I wrong procedure?
details I've attached screenshots.

<img width="698" alt="スクリーンショット 2021-10-13 15 38 55" src="https://user-images.githubusercontent.com/3970110/137080465-e300b3bd-7f34-47a8-a947-3cd454cb9aae.png">
<img width="703" alt="スクリーンショット 2021-10-13 15 38 48" src="https://user-images.githubusercontent.com/3970110/137080485-898e1453-0513-4fde-b277-84bc2fba037a.png">

**Reply by susodapop** [reply]
This is a good catch and something I observed in initial testing as well. Ideally we need to monitor for the `default` and `admin` groups and write to these instead of creating duplicates. I created #74 for this.

**Reply by kaz3284** [reply]
@susodapop I always appreciate your responses!
When do you expect this issue to be addressed? I'd like to know what you expect to do as far as you can tell.
We are currently suspending our internal migration project because we believe this is a critical issue for migration.

**Reply by wtfzambo** [reply]
@kaz3284 yo, I made a pull request to address this issue: #82 . I tested it on a clean v10.0.0 instance and it works properly. If you don't want to wait for merge you can clone my repo here: https://github.com/wtfzambo/redash-toolbelt/tree/fix_duplicate_defaults

**Reply by kaz3284** [reply]
@wtfzambo First of all, thank you for your message and your response!
It would have been enough to know when it was due.
I've seen the PR, and it looks like there are some parts that haven't gone through CI, 
so I'd like to resume our project once they're approved.

**Reply by susodapop** [reply]
Thanks to you both. Will prioritise getting these approved and merged in the next couple days.

**Reply by kaz3284** [reply]
@susodapop Hi. When I tried it in the verification environment at hand, the 「default」 was duplicated in User details as shown in the capture. Is there any problem with this state?
<img width="224" alt="スクリーンショット 2021-10-20 11 53 29" src="https://user-images.githubusercontent.com/3970110/138020728-43ad96b2-d23b-4508-9dcb-a344202ce122.png">

**Reply by susodapop** [reply]
Hi @kaz3284 are you able to reset the destination environment and try again? I ran wtfzambo's subroutine under a few different manual tests before merging and in each case it behaved as expected

(going forward I hope we can create a proper integration test environment for this with a destination instance and a target instance -- that we can get away from manual testing)

**Reply by wtfzambo** [reply]
@kaz3284 was it a completely clean slate environment? Can you provide a bit more extra info about that screenshot? Where exactly was "default" duplicated ?

**Reply by kaz3284** [reply]
@susodapop @wtfzambo thank you for your message.
Migration was executed with only the first admin member.
That phenomenon appears in the detail screen when viewing an individual user in the "User" tab.
Specifically, when you access a URL such as "https://redash.xxxx/users/350".
I hope it's just a display problem, but I'm afraid of inconsistencies.

**Reply by kaz3284** [reply]
If you can't reproduce the same phenomenon, I'll try again tomorrow, JST, in a clean environment.

**Reply by wtfzambo** [reply]
@kaz3284 Currently I'm at home sick, so I don't have my work machine with me at the moment and can't perform a test, but I can formulate an hypothesis on why it's happening:

when creating the admin user on the destination instance, iirc the default group is immediately associated with it. When going through the member list in the `import_groups` function, the groups are associated back to each user present in origin, including admin, but there's no check in place to see if any specific user is already associated with a specific group. Hence, for the group `default`, it seems to get associated twice.

**Reply by susodapop** [reply]
@kaz3284 I think you can check this directly using the metadata database, on the `groups` table. If there are two `default` groups then we have an issue.

**Reply by kaz3284** [reply]
@wtzambo Take care of yourself. Thank you for your response nonetheless! @susodapop Thank you for your message!
We checked the db value in our environment, and it was not a critical issue, with the same groupId set for User as shown in the attachment. I'm glad it's just a minor bug.
<img width="1302" alt="user_table" src="https://user-images.githubusercontent.com/3970110/138200753-e7b16224-517f-4358-8e50-0489b35739a5.png">
![group_table](https://user-images.githubusercontent.com/3970110/138200762-62185e4a-9623-401f-be52-f1c36a318962.png)

**Reply by wtfzambo** [reply]
@kaz3284 so the groups migrate properly, but the users get the default group twice, am I correct?

Btw I didn't even know it was possible to query Redash's own internal metadata from the UI, is it a V10 feature ?

**Reply by kaz3284** [reply]
@wtfzambo That's right. Your interpretation is correct.

By the way, the metadata display is a snapshot of RDS on AWS, converted with Glue and displayed with Athena.

**Reply by wtfzambo** [reply]
@kaz3284 oh alright, that's interesting.

Anyway, I think I'll be able to work on this task not earlier than monday, unless I miracolously feel better tomorrow. It should be relatively easy to address tbh

**Reply by wtfzambo** [reply]
@kaz3284 hello, I made a pull request to address that bug. You can check it [here](https://github.com/getredash/redash-toolbelt/pull/98)
